watchdogd: init at 4.0; add module

vifino opened the pull request:
## Description of changes

https://troglobit.com/projects/watchdogd/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

vifino committed:
watchdogd: init at 4.0

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+  mkPluginOpts = plugin: {
+    enable = mkEnableOption "watchdogd plugin ${plugin}";
+    interval = mkOption {
+      type = types.int;
```
```suggestion
      type = types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.int;
+      default = 300;
+      description = ''
+        The poll interval. Defaults to 300 seconds.
```
I don't think the defaults need to be included in the descriptions. They will be printed in the manual anyway.

```suggestion
        Amount of seconds between every poll.
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.bool;
+      default = false;
+      description = ''
+        Log current stats every poll interval.
+        Defaults to disabled.
```
```suggestion
        Whether to log current stats every poll interval.
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
+    package = mkOption {
+      default = pkgs.watchdogd;
+      defaultText = "pkgs.watchdogd";
+      type = types.package;
+      description = "watchdogd package to use";
+    };
```
```suggestion
    package = mkPackageOptionMD pkgs "wtachdogd" { };
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+    };
+
+    timeout = mkOption {
+      type = types.int;
```
```suggestion
      type = types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.int;
+      default = 15;
+      description = ''
+        The WDT timeout before reset.
+        Defaults to 15 seconds.
```
```suggestion
        The WDT timeout before reset.
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.int;
+      default = 5;
+      description = ''
+        The kick interval, i.e. how often watchdogd(8) should reset the WDT timer.
+        Defaults to 5 seconds.
```
```suggestion
        The kick interval, i.e. how often {manpage}`watchdogd(8)` should reset the WDT timer.
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.bool;
+      default = true;
+      description = ''
+        With safeExit enabled, the daemon will ask the driver to disable the WDT before exiting.
+        However, some WDT drivers (or HW) may not support this
+        Defaults to true.
```
Some of these abbreviations are a bit hard to follow. WDT is probably fine, as it seems to be a shorthand used for this program (assuming WatchDogTimer?), but I think `HW` is a bit unnecessary.

```suggestion
        With {var}`safeExit` enabled, the daemon will ask the driver to disable the WDT before exiting.
        However, some WDT drivers (or hardware) may not support this
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      type = types.lines;
+      default = "";
+      description = ''
+        Additional config to put in watchdogd.conf.
+        See watchdogd.conf(5) for syntax.
```
```suggestion
        Additional config to put in {file}`watchdogd.conf`.
        See {manpage}`watchdogd.conf(5)` for syntax.
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+    systemd.services.watchdogd = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Advanced system & process supervisor";
+      path = [ cfg.package ];
```
Does this program invoke itself assuming it exists in its own `$PATH`? Else I think this line can be removed.

vifino replied:
Do I need to add the package to restartTriggers instead or will that happen automatically?

h7x4 replied:
No that will happen automatically, as the systemd unit itself changes

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+    };
+
+    # Write the config.
+    environment.etc."watchdogd.conf" = {
+      mode = "0644";
+      text = ''
+        timeout = ${toString cfg.timeout}
+        interval = ${toString cfg.interval}
+        safe-exit = ${boolToString cfg.safeExit}
+        ${mkPluginConf "filenr"}
+        ${mkPluginConf "loadavg"}
+        ${mkPluginConf "meminfo"}
+        ${extraConfig}
+      '';
+    };
```
I think it would be better if this was written directly to the nix store using something like `writeFile`, instead of doing indirection via `/etc`.

h7x4 commented on pkgs/by-name/wa/watchdogd/package.nix:
```diff
@@ -0,0 +1,24 @@
+  src = fetchFromGitHub {
+    owner = "troglobit";
+    repo = "watchdogd";
+    rev = "${version}";
```
Unnecessary string interpolation

```suggestion
    rev = version;
```

h7x4 commented on pkgs/by-name/wa/watchdogd/package.nix:
```diff
@@ -0,0 +1,24 @@
+{ lib, stdenv, fetchFromGitHub, pkg-config, autoreconfHook, libite, libuev
+, libconfuse }:
```
This is just a suggestion, but I think it would be nice if these arguments were laid out with newlines in between them. It makes it a lot easier to review with a cleaner diff, if anything were to be added or removed.

```suggestion
{ lib
, stdenv
, fetchFromGitHub
, pkg-config
, autoreconfHook
, libite
, libuev
, libconfuse
}:
```

vifino replied:
I had it this way but nixpkgs-fmt reformatted it again.

h7x4 commented on pkgs/by-name/wa/watchdogd/package.nix:
```diff
@@ -0,0 +1,24 @@
+
+  meta = with lib; {
+    description = "Advanced system & process supervisor for Linux";
+    homepage = "https://troglobit.com/watchdogd.html";
```
Let's add the changelog as well

```suggestion
    homepage = "https://troglobit.com/watchdogd.html";
    changelog = "https://github.com/troglobit/watchdogd/releases/tag/${version}";
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      '';
+    };
+    interval = mkOption {
+      type = types.int;
```
```suggestion
      type = types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+    };
+  };
+
+  mkPluginConf = plugin:
+    let pcfg = cfg.${plugin};
+    in optionalString pcfg.enable ''
+      ${plugin} {
+        enabled  = true
+        interval = ${toString pcfg.interval}
+        logmark  = ${toString pcfg.logmark}
+        warning  = ${toString pcfg.warning}
+        critical = ${toString pcfg.critical}
+      }
+    '';
```
Not a blocker, but this doesn't seem to be used apart from the `config` block. Would you mind moving it over there for better code locality?

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+      '';
+    };
+    warning = mkOption {
+      type = types.int;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.int;
```
I must say I'm not entirely sure what these options are, but I'm assuming they're not supposed to be negative numbers

```suggestion
      type = types.ints.unsigned;
      description = ''
        The high watermark level. Alert sent to log.
      '';
    };
    critical = mkOption {
      type = types.ints.unsigned;
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,128 @@
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
```
This seems a bit unnecessary, considering the `watchdogd` command doesn't seem to have any control unix socket that can be controlled via the cli, or something similar. Are there any cases where a user might want to invoke `watchdogd` from their terminal while it's running in systemd?

vifino replied:
It does, `watchdogctl(1)` connects to `/run/watchdogd.sock`.

h7x4 replied:
Okay! Feel free to mark as resolved :grin:

vifino committed:
nixos/watchdog: add module

vifino commented:
Hey @h7x4, mind reviewing this again and getting this merged?

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,113 @@
+        Whether to log current stats every poll interval.
+      '';
+    };
+    warning = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The critical watermark level. Alert sent to log, followed by reboot or script action.
+      '';
+    };
```
While this might mess up the args of this nice `mkOption` function a bit, I think it would be a lot better user experience for this module if we set some defaults for this. As it stands now, the user would have to set all 6 watermark levels manually, unless I'm reading it wrong. Since the function is never exposed to the user, maybe you could just add two more arguments?

These seem to be the upstream defaults: https://github.com/troglobit/watchdogd/blob/6ae714cd22f24e79270296774e2bfb36570d8209/watchdogd.conf#L92-L165

vifino replied:
I skipped this because of the loadavg stuff - it's based on the core count as far as I can tell.
But I added defaults for the two others with the extra two arguments.

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,113 @@
+  };
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
```
`mkEnableOption` is a bit weird, because it adds `Whether to enable ${desc}.` as the description. In order for this not to become `Whether to enable Advanced system & process supervisor.`, I think we should change it up. How about this?

```suggestion
    enable = mkEnableOption "watchdogd, an advanced system & process supervisor";
```

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,113 @@
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
```
While this is definitely more simple, I think it would be more semantically correct to set it to `nullOr lines`, set the default to `null` and use `optionalString` in `config`. Up to you.

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,113 @@
+    systemd.services.watchdogd = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Advanced system & process supervisor";
+      restartTriggers = [ watchdogdConf ];
```
I do think nixos-rebuild will restart the process whenever the unit changes, and the unit changes anytime this file changes due to the command in `ExecStart`. So this is probably superfluous. Probably not a big deal though.

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,113 @@
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
```
Suggestion: I saw these were mentioned in the descriptions, and I think they are relevant enough to the daemon to be put here.

```suggestion
    systemd.services.watchdogd = {
      documentation = [
        "man:watchdogd(8)"
        "man:watchdogd.conf(5)"
      ];
```

vifino replied:
Huh! What does that do, where does it show up?
Added.

h7x4 replied:
It shows up in (among other things) `systemctl status <service>`. An example would be `systemctl status nix-daemon`, which contains both a manpage and link. It's nice to have when you're staring at a dead systemd unit, as a sort of reminder to go read the docs.

h7x4 commented on pkgs/by-name/wa/watchdogd/package.nix:
```diff
@@ -0,0 +1,32 @@
+, libuev
+, libconfuse
+}:
+stdenv.mkDerivation rec {
+  pname = "watchdogd";
+  version = "3.5";
+
+  src = fetchFromGitHub {
+    owner = "troglobit";
+    repo = "watchdogd";
+    rev = version;
```
nit: consider using the finalAttrs pattern. It's supposed to be better in some ways, but I'm not sure about the specifics.

```suggestion
stdenv.mkDerivation (finalAttrs: {
  pname = "watchdogd";
  version = "3.5";

  src = fetchFromGitHub {
    owner = "troglobit";
    repo = "watchdogd";
    rev = finalAttrs.version;
```

vifino replied:
finalAttrs helps while overriding as it'll update when you override the version, but I think it's not worth it here, since the only relevant place is in the fetcher, where the hash will change.

h7x4 replied:
Sounds fine by me. It's trivial to fix it up, if it ever becomes an issue again or if someone makes a treewide pr.

h7x4 reviewed (commented):
Here's a few more comments

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,121 @@
+      '';
+    };
+    warning = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    } // optionalAttrs (defWarn != null) {
+      default = defWarn;
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
```
On second thought, these two cannot have this type. We even set their defaults to be floats. I'm unsure if they will break without the `.0` suffix on ints.

```suggestion
      type = types.numbers.nonnegative;
      description = ''
        The high watermark level. Alert sent to log.
      '';
    } // optionalAttrs (defWarn != null) {
      default = defWarn;
    };
    critical = mkOption {
      type = types.numbers.nonnegative;
```

vifino replied:
Good call! I completely overlooked the numbers types.
Thanks!

vifino commented:
Hey @h7x4, mind looking at this again?

h7x4 commented:
Hello again!

I spent some time tinkering with this module, to make it more in style with [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). That includes putting most of the config into `settings`, and allowing for freeform settings. I've also added a small test, and marked you as the maintainer, hope that's okay with you?

It did come at the cost of setting some defaults that we discussed earlier not to set. But I think it should be fine, considering these are defaults provided by upstream.

Here's the patches: https://gist.github.com/h7x4/215f4d642ca0b0e67b91354f3b4ada4f

vifino commented:
Hello! :)

I took a look at your patch, it seems very sensible!
Will apply the changes and try to expand the test a bit.

I really appreciate you taking the time to review this!
Your feedback is invaluable!

Thank you!

vifino committed:
nixosTests.watchdogd: init

vifino commented:
Hey @h7x4, the recursive type exploded in the doc building process because of the infinite recursion. I replaced it with a non-recursive variant.

I changed the test so that it now checks if `watchdogctl` can connect to the daemon and runs the builtin tests, so if they pass, it surely works! :)

h7x4 commented on nixos/modules/services/monitoring/watchdogd.nix:
```diff
@@ -0,0 +1,132 @@
+
+    settings = mkOption {
+      type = with types; submodule {
+        freeformType = let
+          valueType = attrsOf (oneOf [
+            bool
+            int
+            float
+            str
+          ]);
+        in attrsOf (oneOf [
+          valueType
+          (attrsOf valueType)
+        ]);
```
This seems wrong. If you're trying to restrict the option to only allow for a single nested layer, I think the `attrsOf` on `valueType` needs to go. Also no need for `oneOf` when there's only two options

```suggestion
        freeformType = let
          valueType = oneOf [
            bool
            int
            float
            str
          ];
        in attrsOf (either valueType (attrsOf valueType));
```

vifino replied:
Yes, that seems more correct, thank you!

vifino commented:
Alright! This sure went up in quality quite a bit! Anything left to do? :)

h7x4 reviewed (approved):
Not as far as I can see :)
